src/Main.cpp

```diff
@@ -1,10 +1,466 @@
+            ws_promise.set_value();
+        }
+    );
+
+    rusty_websocket->onError(
+        [&](string const& error)
+        {
+            cout << "RustySignal WebSocket failed: " << error << endl;
+            ws_promise.set_exception(make_exception_ptr(runtime_error(error)));
+        }
+    );
+
+    rusty_websocket->onClosed([&]() { cout << "RustySignal WebSocket closed" << endl; });
+
+    rusty_websocket->onMessage(
+        [&](variant<binary, string> data)
+        {
+            if (!holds_alternative<string>(data))
+            {
+                return;
+            }
+
+            string error;
+            if (!rusty_decoder.parseJSONMessage(get<string>(data).c_str(), error))
+            {
+                throw invalid_argument(error);
+            }
+
+            string actiontype = rusty_decoder.getActionType();
+
+            if (actiontype == "ping" && client_id_checked)
+            {
+                pong(rusty_websocket, rusty_decoder, local_peer_id);
+            }
+
+            if (signalr_websocket)
+            {
+                if (actiontype == "offer" || actiontype == "answer")
+                {
+                    offerAnswerRustyMessageParser(signalr_websocket, rusty_decoder);
+                }
+                else if (actiontype == "candidate")
+                {
+                    candidateRustyMessageParser(signalr_websocket, rusty_decoder);
+                }
+            }
+        }
+    );
+
+    // wss://signalserverhost?user=yourname
+    const string rusty_url_websocket = rusty_signal_server_host + "?user=" + local_peer_id;
+    cout << "RustySignal WebSocket URL is " << rusty_url_websocket << endl;
+    rusty_websocket->open(rusty_url_websocket);
+
+    while (true)
+    {
+        string leave_session;
+        cout << "Enter with \"leave_session\" to disconnect:" << endl;
+        cin >> leave_session;
+        cin.ignore();
+
+        if (signalr_decoder.checkSessionClosed())
+        {
+            cout << "Session closed: Vehicle server disconnected" << endl;
+            break;
+        }
+
+        if (leave_session != "leave_session")
+        {
+            cout << "Invalid argument: \"leave_session\" to disconnect" << endl;
+        }
+        else
+        {
+            leaveSession(signalr_websocket, signalr_decoder, local_peer_id);
+            break;
+        }
+    }
 
     return 0;
 }
+catch (exception const& error)
+{
+    cout << "Error:" << error.what() << endl;
+    return -1;
+}
+
+void sendInitialPayload(shared_ptr<rtc::WebSocket>& websocket)
+{
+    Json::Value message;
+    message["protocol"] = "json";
+    message["version"] = "1";
+    if (auto ws = make_weak_ptr(websocket).lock())
+    {
+        Json::FastWriter fast;
+        ws->send(fast.write(message));
+    }
+}
+
+void sendSessionCheck(shared_ptr<rtc::WebSocket>& websocket, string& local_peer_id)
+{
+    Json::Value message, content(Json::arrayValue);
+    content.append(local_peer_id);
```

Comment:
Isnt this supposed to be the client_id? Also, I couldnt find out how you would get the client_id by now.

Reply:
The `local_peer_id` is the `client_id`, from the deep_trekker perspective. They are the same.